### ISSUE TITLE
Android export broken in 3.2.4 RC5 - 'apksigner' Invalid keystore format 

### ISSUE THREAD

**Issue by vix597** [godotengine/godot#47043]
**Godot version:**

v3.2.4.rc5.official

**OS/device including version:**

* Windows 10 64-bit Version 20H2 OS Build 19042.867

**Issue description:**

When exporting for Android I get the following error on 3.2.4 RC5 but not on 3.2.3:

```
Failed to load signer "signer #1"
java.io.IOException: Invalid keystore format
        at sun.security.provider.JavaKeyStore.engineLoad(JavaKeyStore.java:663)
        at sun.security.provider.JavaKeyStore$JKS.engineLoad(JavaKeyStore.java:56)
        at sun.security.provider.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:224)
        at sun.security.provider.JavaKeyStore$DualFormatJKS.engineLoad(JavaKeyStore.java:70)
        at java.security.KeyStore.load(KeyStore.java:1445)
        at com.android.apksigner.ApkSignerTool$SignerParams.loadKeyStoreFromFile(ApkSignerTool.java:829)
        at com.android.apksigner.ApkSignerTool$SignerParams.loadPrivateKeyAndCertsFromKeyStore(ApkSignerTool.java:719)
        at com.android.apksigner.ApkSignerTool$SignerParams.loadPrivateKeyAndCerts(ApkSignerTool.java:659)
        at com.android.apksigner.ApkSignerTool$SignerParams.access$500(ApkSignerTool.java:611)
        at com.android.apksigner.ApkSignerTool.sign(ApkSignerTool.java:266)
        at com.android.apksigner.ApkSignerTool.main(ApkSignerTool.java:89)
```

For the same project, this error does not occur with Godot 3.2.3. Both are exporting with debug enabled and both are setup to use the same Java, Android SDK, debug keystore, etc. in the editor settings.

**Steps to reproduce:**

1. Have a system setup to export for Android. See: https://docs.godotengine.org/en/stable/getting_started/workflow/export/exporting_for_android.html
1. Open the attached project with the official Godot 3.2.3
1. Click to install the android export template from the "Project" menu
1. Export using the configured Android export preset. It should succeed.
1. Open the attached project with the official Godot 3.2.4 RC5 and export will trigger the error described above.

**Minimal reproduction project:**

[AndroidExportBug-47043.zip](https://github.com/godotengine/godot/files/6145430/AndroidExportBug-47043.zip)

**Comment by Calinou** [comment]
cc @m4gr3d

**Comment by akien-mga** [comment]
If you generated your `debug.keystore` a while ago, it might not be valid for `apksigner`, see #45341.

Re-creating your `debug.keystore` using the latest instructions from https://docs.godotengine.org/en/latest/tutorials/export/exporting_for_android.html should solve it (namely with `-deststoretype pkcs12`).

**Comment by vix597** [comment]
@akien-mga - I just tried that. I moved my `C:\users\<user>\.android\debug.keystore` to `debug.keystore.old` and then ran:

```
keytool -keyalg RSA -genkeypair -alias androiddebugkey -keypass android -keystore debug.keystore -storepass android -dname "CN=Android Debug,O=Android,C=US" -validity 9999 -deststoretype pkcs12
```

It regenerated the file and I still experience the same crash. I'll upload my debug keystore here if that helps: [debug.keystore.zip](https://github.com/godotengine/godot/files/6168003/debug.keystore.zip)

**Comment by akien-mga** [comment]
What version(s) of the `build-tools` do you have installed in your Android SDK?

**Comment by vix597** [comment]
I went through and basically redid everything. Installed the latest android studio, updated all packages, reinstalled all the dependencies too. I am still seeing the issue. The versions I'm on now after updating everything are below:

* I have Android 10.0 (Q) API Level 29 installed for SDK Platforms (it's the only one I have installed)
* Under SDK Tools I have:
    * Android SDK Build-Tools 31-rc2 (this is the latest I can get)
    * NDK (Side-by-side) (no version listed but it's the latest I could get)
    * CMake (latest available but no version listed)
    * Android Emulator 30.5.3
    * Android SDK Platform-Tools 31.0.1

My editor settings:

* ADB - `C:/Users/<username>/AppData/Local/Android/Sdk/platform-tools/adb.exe`
* jarsigner - `C:/Program Files/AdoptOpenJDK/jdk-8.0.282.8-hotspot/bin/jarsigner.exe`
* Debug keystore - `C:/users/<username>/.android/debug.keystore`

My Windows PATH variable contains: `C:\Program Files\AdoptOpenJDK\jdk-8.0.282.8-hotspot\bin` so I know the `keytool.exe` I'm using is the one that ships with OpenJDK 8.

After updating everything I did the following:
1. Delete my `debug.keystore`
1. Remake it with `keytool -keyalg RSA -genkeypair -alias androiddebugkey -keypass android -keystore debug.keystore -storepass android -dname "CN=Android Debug,O=Android,C=US" -validity 9999 -deststoretype pkcs12`
1. Open Godot 3.2.3 - Test that I can still export for android after all my changes/updates - Export works
1. Open Godot 3.2.4
1. Delete the `android` folder in my game and remake it by running "Install android build template" from the `Project` menu (not sure if I need to do this but wanted to be sure it was generated by Godot 3.2.4)
1. Test export and it still fails with the same error

I noticed the release was changed to Godot 3.3 and that a new 3.3 RC6 is available to download. So I went and grabbed the latest and re-executed the steps above from step 4 to the end but with 3.3 RC6 this time and I see the same "Invalid Keystore format".

**Comment by akien-mga** [comment]
I tested with your `debug.keystore` and it works fine for me. I'm using `apksigner` from Build-Tools 30.0.3 (but I don't know that the 31-rc2 should be incompatible).

Can you try exporting a brand new project with the default export preset, to rule out issues linked to your project's configuration?

**Comment by vix597** [comment]
OKAY! I figured it out lol.

In Android studio when you click to update the build tools it keeps all the old version you have. When I went in I actually had to uncheck the Android build tools to uninstall them. When I clicked apply a dialog came up confirming that I was uninstalling the Android build tools and it listed like 6 versions of the build tools it was going to remove. I click confirm, it uninstalled them, then I checked the box to reinstall and it only installed the one version (which was 31-rc2).

Then I just clicked export again and it worked.

Sorry for the confusion.
